### ISSUE TITLE
chef-client in local mode needs minutes to start processing long run lists

### ISSUE THREAD

**Issue by jwadolowski** [chef/chef#4980]
## Description

I've recently noticed there's something wrong with chef-zero performance. I used to execute chef-zero on my Vagrant boxes and my initial thought was that either Vagrant's chef-zero provisioner or vagrant-berkshelf plugin were responsible for this, but after a bit of debugging it seems to be related to chef-zero itself.
## Chef Version

Chef 12.5.1 and 12.10.24 (the latest version at the time of writing)
## Platform Version

CentOS 6.8
## Replication Case

When chef-zero provisioner is defined in `Vagrantfile`, the following command is executed under the hood:

```
$ /opt/chef/embedded/bin/ruby /usr/bin/chef-client --config /tmp/vagrant-chef/client.rb --json-attributes /tmp/vagrant-chef/dna.json --local-mode --log_level debug --force-formatter
```

This is how `client.rb` and `dna.json` look like in my case:

```
node_name "project-vagrant"
file_cache_path    "/var/chef/cache"
file_backup_path   "/var/chef/backup"
cookbook_path ["/tmp/vagrant-chef/1c314d9dbafcfb313f57c7c0ebb1daa0/cookbooks"]
role_path []
log_level :debug
verbose_logging false
enable_reporting false
encrypted_data_bag_secret nil
chef_zero.enabled true
local_mode true
node_path ["/tmp/vagrant-chef/9a29f1bf43fbc28c28a61610e72eea13/nodes"]
```

```
{
    "run_list": [
        "recipe[project-base::default]",
        "recipe[project-consul::server]",
        "recipe[project-consul::service_dispatcher]",
        "recipe[project-consul::service_te_editor]",
        "recipe[project-consul::service_te_publish]",
        "recipe[project-consul::service_tomcat]",
        "recipe[project-consul::service_solr_master]",
        "recipe[project-consul::service_solr_slave]",
        "recipe[project-consul::service_aem_editor]",
        "recipe[project-consul::service_aem_publish]",
        "recipe[project-consul::service_sftp]",
        "recipe[project-httpd::default]",
        "recipe[project-te::editor]",
        "recipe[project-te::publish]",
        "recipe[project-json-server::default]",
        "recipe[project-tomcat::default]",
        "recipe[project-tomcat::soap_mocks]",
        "recipe[project-solr::master]",
        "recipe[project-solr::slave]",
        "recipe[project-sftp::default]",
        "recipe[project-webapp::editor]",
        "recipe[project-webapp::editor_users]",
        "recipe[project-webapp::editor_agents]",
        "recipe[project-webapp::editor_base]",
        "recipe[project-webapp::editor_zg]",
        "recipe[project-webapp::editor_zg_content]",
        "recipe[project-webapp::editor_tools]",
        "recipe[project-webapp::editor_securecq]",
        "recipe[project-webapp::editor_conf]",
        "recipe[project-webapp::editor_cqsm]",
        "recipe[project-webapp::editor_groovies]",
        "recipe[project-webapp::editor_anatracking]",
        "recipe[project-webapp::editor_grabbit]",
        "recipe[project-webapp::publish]",
        "recipe[project-webapp::publish_users]",
        "recipe[project-webapp::publish_agents]",
        "recipe[project-webapp::publish_base]",
        "recipe[project-webapp::publish_zg]",
        "recipe[project-webapp::publish_zg_content]",
        "recipe[project-webapp::publish_tools]",
        "recipe[project-webapp::publish_conf]",
        "recipe[project-webapp::publish_cqsm]",
        "recipe[project-webapp::publish_groovies]",
        "recipe[project-webapp::publish_anatracking]",
        "recipe[project-webapp::publish_grabbit]"
    ]
}
```

As soon as above command gets executed the following happens (full log attached below):
- chef-zero starts on `http://localhost:8889`
- there are attempts to load some plugins, but all of them fail as there's no such files
- `GET` request to `/nodes/project-vagrant` is sent, which ends with 200 response quite fast
- run list is set
- cookbooks synchronization starts
- `POST` request to `/environments/_default/cookbook_versions` is sent and this is where performance issue starts
- `No chefignore file found at /tmp/vagrant-chef/1c314d9dbafcfb313f57c7c0ebb1daa0/chefignore no files will be ignored` is displayed over 400 times (nothing really happens besides this for ~1.5-3 minutes, varies by Chef run). `/tmp/vagrant-chef/1c314d9dbafcfb313f57c7c0ebb1daa0` is a directory that contains just `cookbooks` subdirectory
- response to `POST /environments/_default/cookbook_versions` comes in and converge starts

It doesn't really matter if I execute `vagrant provision` on my machine or `chef-client` directly inside Vagrant (as presented above) - performance issue is still there. Debug log presence doesn't really matter when it comes to this 1.5-3 minutes slow down.
## Client Output

https://gist.github.com/jwadolowski/155015bfd1a95f6c64e69b2f8979aab0

The log contains also `GET /cookbooks` request, as I wanted to make sure all cookbooks have been successfully uploaded.

**Comment by thommay** [comment]
Has performance got worse, or has it always been this slow - I guess from your note about 12.5 and latest it's been slow for a fair while.

**Comment by jwadolowski** [comment]
It's been slow for a while. I didn't notice that before, as for some time there was a bug in Vagrant that run chef-solo instead of chef-zero no matter what was defined in `Vagrantfile`. This issue was fixed around a year ago (in 1.7.3 or 1.7.4, can't remember exactly).

For the last 6 months (most probably even slightly longer) this slow processing was definitely present on all Vagrant machines I run.

**Comment by lamont-granquist** [comment]
https://github.com/chef/chef-zero/issues/285 is probably the sketch of what needs to be fixed for chef-zero to make it fast
